### PR TITLE
[SPARK-21274][SQL] Implement INTERSECT ALL clause

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Implements INTERSECT ALL clause through query rewrites using existing operators in Spark.  Please refer to [Link](https://drive.google.com/open?id=1nyW0T0b_ajUduQoPgZLAsyHK8s3_dko3ulQuxaLpUXE) for the design.

Input Query
``` SQL
SELECT c1 FROM ut1 INTERSECT ALL SELECT c1 FROM ut2
```
Rewritten Query
```SQL
   SELECT c1
    FROM (
         SELECT replicate_row(min_count, c1)
         FROM (
              SELECT c1,
                     IF (vcol1_cnt > vcol2_cnt, vcol2_cnt, vcol1_cnt) AS min_count
              FROM (
                   SELECT   c1, count(vcol1) as vcol1_cnt, count(vcol2) as vcol2_cnt
                   FROM (
                        SELECT c1, true as vcol1, null as vcol2 FROM ut1
                        UNION ALL
                        SELECT c1, null as vcol1, true as vcol2 FROM ut2
                        ) AS union_all
                   GROUP BY c1
                   HAVING vcol1_cnt >= 1 AND vcol2_cnt >= 1
                  )
              )
          )
```

## How was this patch tested?
Added test cases in SQLQueryTestSuite, DataFrameSuite, SetOperationSuite

### CODE REVIEW

**Comment by viirya** [comment]
Typo in the PR description: `IF (vcol1_cnt > vcol1_cnt, vcol2_cnt, vcol1_cnt)` -> `IF (vcol1_cnt > vcol2_cnt, vcol2_cnt, vcol1_cnt)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1407,6 +1408,87 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
...
+ * Rewritten Query:
+ * {{{
+ *   SELECT c1
+ *   FROM (
+ *        SELECT replicate_row(min_count, c1) AS (min_count, c1)
+ *        FROM (
+ *             SELECT c1,
+ *                    vcol1_cnt,
+ *                    vcol2_cnt,
+ *                    IF (vcol1_cnt > vcol1_cnt, vcol2_cnt, vcol1_cnt) AS min_count
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: typo here `vcol1_cnt > vcol1_cnt` -> `vcol1_cnt > vcol2_cnt`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1407,6 +1408,87 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
...
+ *
+ * Rewritten Query:
+ * {{{
+ *   SELECT c1
+ *   FROM (
+ *        SELECT replicate_row(min_count, c1) AS (min_count, c1)
+ *        FROM (
+ *             SELECT c1,
+ *                    vcol1_cnt,
+ *                    vcol2_cnt,
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Do we need to have `vcol1_cnt` and `vcol2_cnt` here? I think above `replicate_row` only takes `min_count` input.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -1933,6 +1933,22 @@ class Dataset[T] private[sql](
...
+   * Returns a new Dataset containing rows only in both this Dataset and another Dataset while
+   * preserving the duplicates.
+   * This is equivalent to `INTERSECT ALL` in SQL.
+   *
+   * @note Equality checking is performed directly on the encoded representation of the data
+   * and thus is not affected by a custom `equals` function defined on `T`.
+   *
+   * @group typedrel
+   * @since 2.4.0
+   */
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Better to add `resolves columns by position (not by name)`?

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -1475,6 +1475,28 @@ def intersect(self, other):
         """
         return DataFrame(self._jdf.intersect(other._jdf), self.sql_ctx)
 
+    @since(2.4)
+    def intersectAll(self, other):
+        """ Return a new :class:`DataFrame` containing rows in bothe this frame and other frame
```
Comment (viirya) [inline python/pyspark/sql/dataframe.py]: typo: `bothe`. 

frame -> dataframe.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @viirya Thanks !! No we don't. In the actual code, we don't project these columns out. I will fix the doc.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @viirya OK

File: python/pyspark/sql/dataframe.py (hunk 4 above)
Comment (dilipbiswal) [reply python/pyspark/sql/dataframe.py]: @viirya Will fix.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -1407,6 +1408,87 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
...
+ *                    IF (vcol1_cnt > vcol1_cnt, vcol2_cnt, vcol1_cnt) AS min_count
+ *             FROM (
+ *                  SELECT   c1, count(vcol1) as vcol1_cnt, count(vcol2) as vcol2_cnt
+ *                  FROM (
+ *                       SELECT c1, true as vcol1, null as vcol2 FROM ut1
+ *                       UNION ALL
+ *                       SELECT c1, null as vcol1, true as vcol2 FROM ut2
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @viirya OK :-)

**Comment by dilipbiswal** [comment]
@gatorsmile I see this failure in other PRs as well. Is this introduced by some recent changes ?

**Comment by viirya** [comment]
retest this please.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @dilipbiswal  Could you resolve the conflicts? I will start the review after the rebase.

**Comment by dilipbiswal** [comment]
@gatorsmile Rebased.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -1934,6 +1934,23 @@ class Dataset[T] private[sql](
...
+   *
+   * @note Equality checking is performed directly on the encoded representation of the data
+   * and thus is not affected by a custom `equals` function defined on `T`. Also as standard
+   * in SQL, this function resolves columns by position (not by name).
+   *
+   * @group typedrel
+   * @since 2.4.0
+   */
+  def intersectAll(other: Dataset[T]): Dataset[T] = withSetOperator {
+    Intersect(planWithBarrier, other.planWithBarrier, isAll = true)
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @gatorsmile Sure.. how about exceptAll that was checked in today ?

[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: yes. Please do it too.

File: sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+struct<>
+-- !query 7 output
+org.apache.spark.sql.AnalysisException
+IntersectAll can only be performed on tables with the compatible column types. array<int> <> int at the first column of the second table;
+
+
+-- !query 8
+SELECT c1 FROM tab1
+INTERSECT ALL
+SELECT c1, c2 FROM tab2
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out]: use `k` and `v`

File: sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    AS tab1(k, v)
+-- !query 0 schema
+struct<>
+-- !query 0 output
+
+
+
+-- !query 1
+CREATE TEMPORARY VIEW tab2 AS SELECT * FROM VALUES
+    (1, 2),
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out]: also add another duplicate rows for (1, 2);

File: sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out [hunk 9]
```diff
@@ -0,0 +1,212 @@
+-- Automatically generated by SQLQueryTestSuite
+-- Number of queries: 17
+
+
+-- !query 0
+CREATE TEMPORARY VIEW tab1 AS SELECT * FROM VALUES
+    (1, 2), 
+    (1, 2),
+    (1, 3),
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out]: also add another duplicate row (1, 3)

File: sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+-- !query 0 schema
+struct<>
+-- !query 0 output
+
+
+
+-- !query 1
+CREATE TEMPORARY VIEW tab2 AS SELECT * FROM VALUES
+    (1, 2), 
+    (2, 3)
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out]: add one more row (3, 4)

**Comment by gatorsmile** [comment]
The code looks good to me. Let us improve the test cases.

**Comment by dilipbiswal** [comment]
@gatorsmile Thank you.. I will make the changes.

File: sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+INTERSECT ALL
+SELECT * FROM tab2
+-- !query 10 schema
+struct<k:int,v:int>
+-- !query 10 output
+1	2
+1	2
+2	3
+NULL	NULL
+NULL	NULL
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/intersect-all.sql.out]: This misses one row (1, 3). Could you investigate the cause?

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 12]
```diff
@@ -164,7 +164,12 @@ object SetOperation {
   def unapply(p: SetOperation): Option[(LogicalPlan, LogicalPlan)] = Some((p.left, p.right))
 }
 
-case class Intersect(left: LogicalPlan, right: LogicalPlan) extends SetOperation(left, right) {
+case class Intersect(
+   left: LogicalPlan,
+   right: LogicalPlan,
+   isAll: Boolean = false) extends SetOperation(left, right) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: not a big deal at all but this has tree spaces ..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 13]
```diff
@@ -529,9 +529,13 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
         throw new IllegalStateException(
           "logical distinct operator should have been replaced by aggregate in the optimizer")
-      case logical.Intersect(left, right) =>
+      case logical.Intersect(left, right, false) =>
         throw new IllegalStateException(
-          "logical intersect operator should have been replaced by semi-join in the optimizer")
+          "logical intersect  operator should have been replaced by semi-join in the optimizer")
+      case logical.Intersect(left, right, true) =>
+        throw new IllegalStateException(
+          "logical intersect operator should have been replaced by union, aggregate" +
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: nit: looks we need a space for ` aggregate"` -> ` aggregate "`

**Comment by gatorsmile** [comment]
@dilipbiswal Please address the style issues in your other PRs.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by gatorsmile** [comment]
@dilipbiswal The merged PR does not pick up your last commit.

**Comment by dilipbiswal** [comment]
@gatorsmile Ok Sean..

**Pull request closed without merging**